k8s.gcr.io/Vanity-Domain-Flip.md

```diff
@@ -0,0 +1,191 @@
+# The Problem
+
+Currently, `k8s.gcr.io` is a vanity domain that points to
+`gcr.io/google-containers` (Google-owned and managed). This is a problem because
+`k8s.gcr.io` is an alias used throughut the Kubernetes codebase. As Kubernetes
+is a community-owned project, `k8s.gcr.io` should instead point to a
+community-controlled repo.
+
+# The Solution
+
+The community has created a new repo called `gcr.io/k8s-artifacts-prod`, and it
+has been agreed that the community should use it as the new place to push
+production images (instead of `gcr.io/google-containers`). **We can solve the
+above problem by flipping the vanity domain (`k8s.gcr.io`) from
+`gcr.io/google-containers` to `gcr.io/k8s-artifacts-prod`**. This way, no change
+needs to be made in the Kubernetes codebase.
+
+The minimum prerequisite is that the existing images in `google-containers` must
+be copied into `k8s-artifacts-prod` in order to ensure that the domain flip
+happens transparently without incurring any interruptions. However there are
+other infrastructural improvements that the community has designed, such as
+explicit backups, disaster recovery, and also auditing and alerting.
+
+The rest of this document explains the infrastructural improvements surrounding
+`gcr.io/k8s-artifacts-prod`.
+
+# How New Images Get Promoted (Pushed) to Production
+
+To get new images into `gcr.io/google-containers`, a Googler must approve a
+change in Google's private repository.
+
+On the other hand, `gcr.io/k8s-artifacts-prod` is integrated with a
+publicly-visible GitHub repository, named [k8s.io][k8sio]. The [promoter][CIP]
+watches this repository for changes and promotes images. In addition, a system
+of setting up staging repos, and promoting from them into
+`gcr.io/k8s-artifacts-prod` has been [created][staging-subproject] so that owners of
+subprojects in the community can take control of how their images are released.
+
+## The Promoter (cip)
+
+The [Container Image Promoter][CIP] (henceforth "the promoter") is the OSS
+rewrite of the [promoter used internally within Google][internal-promoter]. It
+works by reading in a set of promoter manifests (YAMLs) that describe the
+desired state of a Docker registry's image contents, and proceeds to copy in any
+missing images. Currently the toplevel `k8s.gcr.io` directory at the [k8s.io
+Github repo][k8sio] defines such a set of promoter manifests.
+
+## Prow Integration
+
+The act of invoking the promoter as a postsubmit against the k8s.io repo is done
+by [Prow][prow], as the `post-k8sio-cip` Prow job. There are other Prow jobs
+that integrate with the promoter, and the ones relevant to this doc are outlined
+in the table below:
+
+- [`post-k8sio-cip`](https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/test-infra/test-infra-trusted.yaml)
```

Comment:
Is there a publicly visible log of Prow's actions?  We should be able to see the history of this stuff.

@Katharine ?

Reply:
https://k8s-testgrid.appspot.com/sig-release-misc#post-k8sio-cip and https://prow.k8s.io/?job=post-k8sio-cip (this one has less history though, so probably not the best link). I will edit all the descriptions to have a link to testgrid as well.

Reply:
The best way to see the longest available history of jobs is probably https://prow.k8s.io/job-history/kubernetes-jenkins/logs/post-k8sio-cip - which will give you every run of a job within our log retention period (which is currently ninety days).

Reply:
Cool.

Linus those are some verbose logs!  Also, looking at logs, you have a bug:

```
I0211 14:20:42.242882      12 inventory.go:466] edge {{gcr.io/k8s-staging-capi-kubeadm   %!s(bool=true)} {cluster-api-kubeadm-controller v0.1.4} sha256:616ebe071066449bef90701cef646f568af64f78a8997467a4ef0db8030557b9 {eu.gcr.io/k8s-artifacts-prod/capi-kubeadm k8s-infra-gcr-promoter@k8s-artifacts-prod.iam.gserviceaccount.com  %!s(bool=false)} {cluster-api-kubeadm-controller v0.1.4}}: skipping because it was already promoted (case 1)
```

Note `%!s(bool=false)`

Reply:
Linus - can you add the link to the job history to this doc, please

Reply:
Ack